src/core/server/server.cc

```diff
@@ -84,6 +86,272 @@
+    if (security_connector == nullptr) {
+      // Unable to create secure server with credentials
+      return absl::nullopt;
+    }
+    new_args = (*args_result).SetObject(security_connector);
+  }
+  MutexLock lock(&mu_);
+  // Since we let go of the lock earlier, we need to protect ourselves against
+  // time-of-check-to-time-of-use cases. The server may have stopped serving
+  // or the connection manager may have changed before we add the connection
+  // to the list of tracked connections.
+  if (!is_serving_ || connection_manager != connection_manager_) {
+    // Not serving
+    return absl::nullopt;
+  }
+  connections_.emplace(std::move(connection));
+  return new_args;
+}
+
+void Server::ListenerState::OnHandshakeDone(
+    ListenerInterface::LogicalConnection* connection) {
+  if (server_->config_fetcher() != nullptr) {
+    return;
+  }
+  // There is no config fetcher, so we can remove this connection from being
+  // tracked immediately.
+  OrphanablePtr<ListenerInterface::LogicalConnection> connection_to_remove;
+  {
+    // Remove the connection if it wasn't already removed.
+    MutexLock lock(&mu_);
+    auto connection_handle = connections_.extract(connection);
+    if (!connection_handle.empty()) {
+      connection_to_remove = std::move(connection_handle.value());
+    }
+    // We do not need to check connections_to_be_drained_list_ since that only
+    // gets set if there is a config fetcher.
+  }
+}
+
+void Server::ListenerState::RemoveLogicalConnection(
+    ListenerInterface::LogicalConnection* connection) {
+  if (server_->config_fetcher() == nullptr) {
+    // The connection was removed from being tracked in `OnHandshakeDone`.
+    return;
+  }
+  OrphanablePtr<ListenerInterface::LogicalConnection> connection_to_remove;
+  {
+    // Remove the connection if it wasn't already removed.
+    MutexLock lock(&mu_);
+    auto connection_handle = connections_.extract(connection);
+    if (!connection_handle.empty()) {
+      connection_to_remove = std::move(connection_handle.value());
+      return;
+    }
+    // The connection might be getting drained.
+    for (auto it = connections_to_be_drained_list_.begin();
+         it != connections_to_be_drained_list_.end(); ++it) {
+      auto connection_handle = it->connections.extract(connection);
+      if (!connection_handle.empty()) {
+        connection_to_remove = std::move(connection_handle.value());
+        if (it->connections.empty()) {
+          // Cancel the timer if the set of connections is now empty.
+          if (event_engine()->Cancel(drain_grace_timer_handle_)) {
+            // Only remove the entry from the list if the cancellation was
+            // actually successful. OnDrainGraceTimer() will remove if
+            // cancellation is not successful.
+            connections_to_be_drained_list_.erase(it);
+            MaybeStartNewGraceTimerLocked();
+          }
+        }
+        return;
```

Comment:
This return is very deeply nested and it's hard to match up with which logic it's corresponding to... is there a way to make this a little bit clearer? - maybe moving the removal action to a separate function so that the body of the if gets down to just `{ WhateverAction(); return; }`